In [1]:
import os
if 'jbook' in os.getcwd():
    os.chdir(os.path.abspath(os.path.join("../..")))
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

In [2]:
from importlib import reload  # Not needed in Python 2
import logging
reload(logging)
logging.basicConfig(format='%(asctime)s %(levelname)s:%(message)s', level=logging.DEBUG, datefmt='%I:%M:%S')

# Data Quality Assessment
The second stage of data processing is the Data Quality Assessment. This stage ensures that our dataset is ready for subsequent analysis and modeling tasks. By identifying and rectifying data quality issues early, we can avoid potential pitfalls that might compromise the integrity and accuracy of our results.

## Data Quality Checks
In this stage, we employ a series of tasks designed to identify and address any noise or irregularities within the dataset. Each task focuses on a specific aspect of data quality, ranging from detecting duplicate entries to identifying profanity, special patterns, and other potential sources of bias or distortion.
1. **Duplicate Rows**: We identify and remove duplicate entries to ensure that each observation is unique, preventing skewed analyses and inflated metrics.
2. **Null Values**: We detect and handle missing data appropriately, which could involve imputation, deletion, or flagging incomplete records for further investigation.
3. **Outliers**: Check for outliers in numeric columns using the non-parametric Interquartile Range (IQR) method.
4. **Non-English Text**: We check for and address non-English text in reviews and app names, as they may not be relevant to our analysis or could require special handling.
5. **Emojis**: Emojis can carry significant meaning in certain contexts but might also introduce noise. We identify and decide on their treatment—whether to retain, remove, or translate them into textual representations.
6. **Excessive Special Characters**: Special characters can disrupt text analysis and need to be managed, either by cleaning or encoding them appropriately.
7. **Invalid Dates**: We verify that date values fall within expected ranges and formats, correcting or flagging anomalies for further review.
8. **Invalid Ratings**: Ratings that fall outside the expected scale (e.g., 1 to 5) are identified and corrected or flagged.
9. **Profanity**: We detect and handle profane content to ensure that our dataset adheres to appropriate usage standards, especially if it's intended for public or sensitive applications.
10. **Special Patterns**: We identify and manage special patterns such as URLs, phone numbers, and emails. These patterns could be indicative of spam or need to be anonymized to protect privacy.

By conducting these data quality checks, we ensure that our dataset is clean, reliable, and ready for detailed analysis. This foundational step sets the stage for accurate insights and robust conclusions in the subsequent phases of our data processing pipeline.

In [3]:

from appinsight.data_prep.dqa import DataQualityAssessment, DQAConfig

04:58:02 INFO:Note: NumExpr detected 24 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
04:58:02 INFO:NumExpr defaulting to 8 threads.


We've encapsulated the data quality assessment process in a `DataQualityAssessment` class. Configured with source and target files, this class conducts the 10 data quality checks, marking the observations that require attention.

In [4]:
config = DQAConfig(force=True)
warnings.filterwarnings("ignore", category=FutureWarning)
dqa = DataQualityAssessment(config=config)
data = dqa.execute()

04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Checking if endpoint exists.
04:58:03 DEBUG:Creating pipeline
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58:03 DEBUG:Instantiated the dqa cache manager.
04:58



# ============================================================================ #
#                        DataQualityAssessment Pipeline                        #
# ============================================================================ #

Task ReadTask completed successfully. Runtime: 00 Minutes 00.482469 Seconds


04:58:05 DEBUG:Cache for dqa_is_duplicate already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_is_duplicate_review_id already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_has_null_values already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_vote_sum_outlier already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_vote_count_outlier already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_has_emojis already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_has_excessive_special_chars already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_has_invalid_date already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_has_invalid_rating already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_contains_email already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_contains_url already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_contains_phone_number already exists. Skipping task.
04:58:05 DEBUG:Cache for dqa_has_non_english_review already exists.

Task DQAInitTask completed successfully. Runtime: 00 Minutes 00.969415 Seconds
Task DetectDuplicateRowTask completed successfully. Runtime: 00 Minutes 00.003864 Seconds
Task DetectDuplicateRowTask completed successfully. Runtime: 00 Minutes 00.002504 Seconds
Task DetectNullValuesTask completed successfully. Runtime: 00 Minutes 00.002175 Seconds
Task DetectOutliersTask completed successfully. Runtime: 00 Minutes 00.002124 Seconds
Task DetectOutliersTask completed successfully. Runtime: 00 Minutes 00.002083 Seconds
Task DetectEmojiTask completed successfully. Runtime: 00 Minutes 00.001888 Seconds
Task DetectSpecialCharacterTask completed successfully. Runtime: 00 Minutes 00.002324 Seconds
Task DetectInvalidDatesTask completed successfully. Runtime: 00 Minutes 00.002663 Seconds
Task DetectInvalidRatingsTask completed successfully. Runtime: 00 Minutes 00.001605 Seconds
Task DetectEmailTask completed successfully. Runtime: 00 Minutes 00.002068 Seconds
Task DetectURLTask completed successful

04:58:05 DEBUG:Data now has 13 columns.
04:58:05 DEBUG:Obtaining dqa_is_duplicate_review_id from cache with 221666 rows.
04:58:05 DEBUG:Data now has 14 columns.
04:58:05 DEBUG:Obtaining dqa_has_null_values from cache with 221666 rows.
04:58:05 DEBUG:Data now has 15 columns.
04:58:05 DEBUG:Obtaining dqa_vote_sum_outlier from cache with 221666 rows.
04:58:06 DEBUG:Data now has 16 columns.
04:58:06 DEBUG:Obtaining dqa_vote_count_outlier from cache with 221666 rows.
04:58:06 DEBUG:Data now has 17 columns.
04:58:06 DEBUG:Obtaining dqa_has_emojis from cache with 221666 rows.
04:58:06 DEBUG:Data now has 18 columns.
04:58:06 DEBUG:Obtaining dqa_has_excessive_special_chars from cache with 221666 rows.
04:58:07 DEBUG:Data now has 19 columns.
04:58:07 DEBUG:Obtaining dqa_has_invalid_date from cache with 221666 rows.
04:58:07 DEBUG:Data now has 20 columns.
04:58:07 DEBUG:Obtaining dqa_has_invalid_rating from cache with 221666 rows.
04:58:07 DEBUG:Data now has 21 columns.
04:58:07 DEBUG:Obtaining d

Task DQAFinalizeTask completed successfully. Runtime: 00 Minutes 04.271583 Seconds
Task WriteTask completed successfully. Runtime: 00 Minutes 00.799485 Seconds


                             DataQualityAssessment                              
                          Pipeline Start | 2024-06-29 16:58:03.562412
                           Pipeline Stop | 2024-06-29 16:58:10.165033
                        Pipeline Runtime | 00 Minutes 06.602621 Seconds




# ============================================================================ #



Let's get a summary of the data quality issues by type.

In [5]:
dqa.overview()

04:58:10 DEBUG:matplotlib data path: /home/john/miniconda3/envs/appinsight/lib/python3.10/site-packages/matplotlib/mpl-data
04:58:10 DEBUG:CONFIGDIR=/home/john/.config/matplotlib
04:58:10 DEBUG:interactive is False
04:58:10 DEBUG:platform is linux
04:58:10 DEBUG:CACHEDIR=/home/john/.cache/matplotlib
04:58:10 DEBUG:Using fontManager instance from /home/john/.cache/matplotlib/fontlist-v330.json


## Data Quality Impressions

The data quality assessment (DQA) conducted on the AppVoC dataset revealed several key issues that need to be addressed to ensure the integrity and reliability of the analysis. These issues include:

- **Duplicates**: A small percentage (<1%) have duplicate review ids.
- **Null Values**: Fortunately, there were no null values detected in the dataset.
- **Invalid Entries**: There were no invalid dates or invalid ratings found.
- **Outliers**: Outliers were identified in vote sums, and vote counts which *could* potentially distort analysis results.
- **Non-English**: A notable proportion of the app names (15%) and review content (4.2%) was flagged for being non-English.
- **Special Characters**: A small percentage of reviews were noted for the presence of special characters in excessive proportions.
- **Profanity & Sensitive Information**: Instances of profanity and the presence of email addresses, URLs, or phone numbers were minimal.

Given these findings, the next step is to identify and treat the high-impact data quality issues. This may involve, without limitation:

- **Removing Duplicates**: Eliminating observations with duplicate review IDs.
- **Handling Outliers**: Identifying and appropriately managing outliers in vote sums, and vote counts.
- **Addressing Non-English Text**: Filtering or translating non-English reviews. 
- **filtering Noise**: Filtering or removing excessive special characters from reviews.
- **Ensuring Clean Content**: Censor or remove reviews containing profanity, and personal identifying information such as phone numbers, URLs, and email addresses.

Cue the action!